### PR TITLE
Add data lab

### PR DESCRIPTION
**Pull request by RazvanN7**
Add initial files for the data lab

### CODE REVIEW

File: content/chapters/data/lab/README.md [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+The C implementation manages the memory manually.
+You can observe that all allocations are performed via `malloc` and the memory is freed using `free`.
+Arrays can be defined as static (on the stack) or dynamic (a pointer to some heap memory).
+Stack memory need not be freed, hence static arrays are automatically de-allocated.
+Heap memory, however, is managed by the user, therefore it is the burden of the programmer to find the optimal memory strategy.
+This offers the advantage that you can fine tune the memory usage depending on your application, but this comes with a cost: more often than not, managing memory is a highly complex error-prone task.
+
+#### Python implementation
+
+The python implementation of the program has no notion of memory allocation.
```
Comment (costinsin) [inline content/chapters/data/lab/README.md:59]:
```suggestion
The Python implementation of the program has no notion of memory allocation.
```

File: content/chapters/data/lab/README.md [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+It simply defines variables and the garbage collector takes care of allocating and deallocating memory.
+Notice how the destructor is called automatically at some point when the garbage collector deems that the list is not used anymore.
+Garbage collection lifts the burden of memory management from the user, however, it may be unsuitable for certain scenarios.
+For example, real-time applications that need to take action immediately once a certain event occurs cannot use a garbage collector. That is because the GC usually stops the application to free dead objects.
+
+#### D implementation
+
+The previous 2 examples have showcased extreme situations: full manual vs. full automatic memory management.
+In D, both worlds are combined: variables may be allocated manually on the stack/heap or allocated via the garbage collector (for brevity, malloc allocation is not presented in this example).
+Arrays that are allocated on the stack behave the same as in C, whereas array allocated with the garbage collector mimic python lists. Classes are also garbage collected.
```
Comment (costinsin) [inline content/chapters/data/lab/README.md:69]:
```suggestion
Arrays that are allocated on the stack behave the same as in C, whereas array allocated with the garbage collector mimic Python lists. Classes are also garbage collected.
```

File: content/chapters/data/lab/README.md [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+In this scenario some random memory is overwritten with `5`.
+The third example exhibits a manifestation of the previous design flaw where the return address of the `main` function is overwritten with `0`, thus leading to a segmentation fault.
+
+Although today it seems obvious that such behavior should not be accepted, we should take into account that the context in which the C language was created was entirely different than today.
+At that time the resource constraints - DRAM memory was around a few KBs, operating systems were at their infancy, branch predictors did not exist etc. - were overwhelming.
+Moreover, security was not a concern because the internet basically did not exist. As a consequence, the language was not developed with memory safety in mind.
+
+#### Python implementation
+
+Technically, it is not possible to do any memory corruption in python (that is if you avoid calling C functions from it).
```
Comment (costinsin) [inline content/chapters/data/lab/README.md:95]:
```suggestion
Technically, it is not possible to do any memory corruption in Python (that is if you avoid calling C functions from it).
```

File: content/chapters/data/lab/README.md [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+Pointers do not formally exist and any kind of array access is checked to be within its bounds.
+The example simply showcases what happens when an out of bounds access is performed - an IndexError is thrown and execution halts.
+
+#### D implementation
+
+The D implementation uses almost the same code as the C implementation, but suffers from minor syntax modidications.
+In essence, the two implement the same logic.
+When compiling this code, it can be observed that the D compiler notices at compile time that an out of bounds access is performed.
+This makes sense, since a static array cannot modify its length and therefore the compiler has all the information to spot the mistake.
+The only way to make the code compile si to comment the faulting lines or to replace the out of bounds index with a corect one.
```
Comment (costinsin) [inline content/chapters/data/lab/README.md:105]:
```suggestion
The only way to make the code compile is to comment the faulting lines or to replace the out of bounds index with a correct one.
```

File: content/chapters/data/lab/README.md [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+
+For this exercise you will need to identify the programming mistake that makes it possible to corrupt memory.
+Navigate to the `memory-corruption` folder.
+Inspect the source file `segfault.c`.
+
+1. What does the program do? (this could be a quiz in the final form)
+2. Compile and run it. What happens?
+3. Debug the program and find the line that causes the segfault. (Note: although using printfs is a viable option, we strongly suggest you use gdb)
+4. Fix the program.
+5. Analyze the corresponding python and D implementation.
```
Comment (costinsin) [inline content/chapters/data/lab/README.md:129]:
```suggestion
5. Analyze the corresponding Python and D implementation.
```

File: content/chapters/data/lab/README.md [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+
+### Memory corruption
+
+For this exercise you will need to identify the programming mistake that makes it possible to corrupt memory.
+Navigate to the `memory-corruption` folder.
+Inspect the source file `segfault.c`.
+
+1. What does the program do? (this could be a quiz in the final form)
+2. Compile and run it. What happens?
+3. Debug the program and find the line that causes the segfault. (Note: although using printfs is a viable option, we strongly suggest you use gdb)
```
Comment (costinsin) [inline content/chapters/data/lab/README.md:127]:
```suggestion
3. Debug the program and find the line that causes the segfault. (Note: although using `printf`s is a viable option, we strongly suggest you use `gdb`)
```

File: content/chapters/data/lab/README.md [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+
+## Strategies on memory allocation
+### Manual memory management
+### Automatic memory management
+### Hybrid approach
+
+
+Although the OS
+
+To better manage a programs memory, the operating systems creates an address space for each process.
```
Comment (alexandruradovici) [inline content/chapters/data/lab/README.md:26]:
```suggestion
To better manage a program's memory, the operating systems creates an address space for each process.
```

**Review by alexandruradovici — CHANGES_REQUESTED** [review]

File: content/chapters/data/lab/content/working-memory.md [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+1. Try to modify the `ca`,`cp` and `cp2` variables by assigning some other value to them.
+   Explain the behavior.
+
+#### Quiz
+
+[Quiz](quiz/memory-access.md)
+
+### Memory Protection
+
+Memory contents (both code an data) is separated into sections or zones.
```
Comment (RazvanN7) [inline content/chapters/data/lab/content/working-memory.md:53]:
```suggestion
Memory contents (both code and data) are separated into sections or zones.
```

File: content/chapters/data/lab/content/working-memory.md [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+Inspect the `mem_access.c` source file.
+
+1. Describe each variable by completing its **(address, size, access rights)** tuple.
+
+1. Try to modify the `ca`,`cp` and `cp2` variables by assigning some other value to them.
+   Explain the behavior.
+
+#### Quiz
+
+[Quiz](quiz/memory-access.md)
```
Comment (Alex-deVis) [inline content/chapters/data/lab/content/working-memory.md:49]:
```suggestion
[Quiz](../quiz/memory-access.md)
```

File: content/chapters/data/lab/content/process-memory.md [hunk 10]
```diff
@@ -0,0 +1,606 @@
...
+893333   20996    7128  921457   e0f71 hello-static
+
+$ size hello-dynamic
+text    data     bss     dec     hex filename
+4598     736     824    6158    180e hello-dynamic
+```
+
+#### Quiz
+
+Based on the information above, answer [this quiz](quiz/static-dynamic.md).
```
Comment (Alex-deVis) [inline content/chapters/data/lab/content/process-memory.md:209]:
```suggestion
Based on the information above, answer [this quiz](../quiz/static-dynamic.md).
```

**Review by Alex-deVis — COMMENTED** [review]
Update path for quizzes

File: content/chapters/data/lab/content/investigate-memory.md [hunk 11]
```diff
@@ -0,0 +1,510 @@
...
+[jemalloc](http://jemalloc.net/) is a featureful allocator that is intended to replace the standard allocator in the standard C library (libc).
+jemalloc provides replacements for the general `malloc()` and `free()` functions and also provides a custom API targeted for performance tuning.
+
+As [documented](https://github.com/jemalloc/jemalloc/wiki/Getting-Started) there are multiple ways to use `jemalloc`, the easiest of which is to use the `LD_PRELOAD` environment variable and preload the library and hook into `malloc()` and `free()` function calls.
+
+First install `jemalloc` on our system.
+On your typical Ubuntu / Debian-based system, use `apt`:
+
+```
+student@os:~/.../data/lab/content$ sudo apt -y install libjemalloc-dev libjemalloc1
```
Comment (wall-hacker) [inline content/chapters/data/lab/content/investigate-memory.md:245]: libjemalloc1 may not be present on some systems. it appears to be called libjemalloc2 on some. libjemalloc-dev should provide libjemalloc[12] as a dependency.

File: content/chapters/data/lab/content/memory-support.md [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+
+**Manual memory management** (MMM) is one of the most difficult tasks.
+Even experienced programmers make mistakes when tackling such a complicated endeavor.
+As a consequence, the programming world has been migrating towards languages that offer automatic memory management (AMM).
+AMM programming languages typically offer a garbage collector that tracks down the usage of objects and frees memory once no references exist to a given object.
+As a consequence garbage collected programming languages are easier to use and safer.
+However, this comes with a cost: the garbage collector, in most cases, requires a significant amount of resources to run.
+Therefore, for performance-critical systems, MMM is still the preferred solution.
+
+A middle-ground between programming languages that have AMM (Java, Python, Rust, D) and those that do not (C, C++) is represented by those languages that do not have builtin AMM but offer the possibility to implement it as a library solution (C++, D).
```
Comment (costinsin) [inline content/chapters/data/lab/content/memory-support.md:11]:
```suggestion
A middle-ground between programming languages that have AMM (Java, Python, D) and those that do not (C, C++, Rust) is represented by those languages that do not have builtin AMM but offer the possibility to implement it as a library solution (C++, D, Rust).
```

Rust does not have AMM [1]. But it has a reference counting structure in the std library [2].

[1] https://stackoverflow.com/questions/40658045/does-rusts-memory-management-result-in-fragmented-memory#:~:text=Rust%20does%20not%20have%20automatic,going%20on%20behind%20the%20scenes.
[2] https://doc.rust-lang.org/stable/std/rc/struct.Rc.html

**Review by costinsin — CHANGES_REQUESTED** [review]

File: content/chapters/data/lab/content/memory-support.md (hunk 12 above)
Comment (RazvanN7) [reply content/chapters/data/lab/content/memory-support.md:11]: Sorry, i had Swift in mind, but typed Rust ...

File: .gitignore [hunk 13]
```diff
@@ -1,5 +1,8 @@
+<<<<<<< HEAD
 # Prerequisites
 
+=======
+>>>>>>> d0b3c34... data/lab: Introduce Data lab
```
Comment (teodutu) [inline .gitignore:5]: Remove these conflicts. IIRC the intention was to remove `*.d` files from here.

**Review by teodutu — APPROVED** [review]
On top of the TODOs that need to be solved, some exercises, such as `reference-counting/refcount_skel.d` also need solutions. But this is mergeable as it is IMO.

**Comment by RazvanN7** [comment]
@teodutu sorry, I missed your comment. I think this is ready now.

**Comment by Alex-deVis** [comment]
@RazvanN7, I think we should solve conflicts and merge this so people will have all information in one place before the winter break.

**Review by razvand — APPROVED** [review]
Works, thanks

**Pull request merged**